# Bayesian Optimization


In [1]:
!pip install hpsklearn

## Load data

In [1]:
import numpy as np
import pandas as pd
data_aia = pd.read_csv('AIA_allfeature.csv')
data_train = pd.read_csv('AIA_train.csv')
data_test = pd.read_csv('AIA_test.csv')

origin_cate_feature = ['service_type', 'complaint_level', 'contract_type', 'gender', 'is_mix_service',
                       'is_promise_low_consume',
                       'many_over_bill', 'net_service']

origin_num_feature = ['1_total_fee', '2_total_fee', '3_total_fee', '4_total_fee',
                      'age', 'contract_time',
                      'former_complaint_fee', 'former_complaint_num',
                      'last_month_traffic', 'local_caller_time', 'local_trafffic_month', 'month_traffic',
                      'online_time', 'pay_num', 'pay_times', 'service1_caller_time', 'service2_caller_time']
feature = data_aia.columns[:-2]
data_aia[origin_cate_feature] = data_aia[origin_cate_feature].astype('category')
data_train[origin_cate_feature] = data_train[origin_cate_feature].astype('category')
data_test[origin_cate_feature] = data_test[origin_cate_feature].astype('category')

In [2]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

data_train['label']=le.fit_transform(data_train['label'].values)
data_train

service_type complaint_level contract_type gender is_mix_service  \
0                 4               0             3      2              0   
1                 1               0             0      2              0   
2                 1               0             0      2              0   
3                 1               0             0      1              0   
4                 1               0             0      1              0   
...             ...             ...           ...    ...            ...   
767520            1               0             0      2              0   
767521            1               0             0      2              0   
767522            4               0             1      1              0   
767523            1               0             0      2              0   
767524            4               1             1      1              1   

       is_promise_low_consume many_over_bill net_service  1_total_fee  \
0                           0              0           4        78.18   
1                           0              1           4        86.60   
2                           0              1           4        43.80   
3                           0              1           4        73.50   
4                           0              1           4        74.00   
...                       ...            ...         ...          ...   
767520                      0              1           4        24.70   
767521                      0              1           4        52.70   
767522                      1              0           4       240.00   
767523                      0              1           4        39.40   
767524                      1              0           4       106.00   

        2_total_fee  ...  4_total_feeW1  4_total_feeW2  4_total_feeW3  \
0              56.0  ...       1.926587       3.013871      -0.752324   
1              52.8  ...       1.999853       3.047367      -0.361521   
2              54.1  ...       2.310467       3.029250      -0.423323   
3              57.9  ...      -0.422283       2.799761      -2.208144   
4              74.8  ...       0.171605      -0.095337      -0.001438   
...             ...  ...            ...            ...            ...   
767520         34.6  ...      -0.064877       1.604699      -0.634678   
767521         55.5  ...       2.022993       2.439579      -0.937087   
767522        240.0  ...       3.052548       0.365216       3.289185   
767523         34.1  ...       1.775873       1.997064      -1.009133   
767524        106.0  ...      -2.156572       2.710832       0.357581   

        4_total_feeW4  4_total_feeW5  4_total_feeW6  4_total_feeW7  \
0            1.272943       1.895517       1.827070      -0.852203   
1            0.384707       1.156852      -1.060000       1.632733   
2            0.249131       1.706402      -0.829686       1.532641   
3            0.369489       2.037206       1.006464       0.410433   
4           -0.208894       0.079483       0.242796      -0.143514   
...               ...            ...            ...            ...   
767520       0.524983       0.293704      -0.723036       0.589189   
767521      -0.007577       2.429035      -1.171271       1.070634   
767522      -1.867787      -0.353546       0.008393       2.455316   
767523      -0.048375       3.303548      -0.749326       0.001213   
767524      -1.293669       2.231160      -0.190362      -1.633388   

        4_total_feeW8  4_total_feeW9  label  
0           -1.497426      -2.037968      1  
1           -3.290275       1.833175      0  
2           -2.907971       1.440202      4  
3           -2.871569       0.674421      4  
4           -0.651352      -0.096339      4  
...               ...            ...    ...  
767520      -4.733323       2.983706      4  
767521      -2.542449       0.990967      4  
767522      -0.775706       1.715284     10  
767523      -2.243190       1.000716      4  
76752

In [3]:
X = data_train[feature]
X

service_type complaint_level contract_type gender is_mix_service  \
0                 4               0             3      2              0   
1                 1               0             0      2              0   
2                 1               0             0      2              0   
3                 1               0             0      1              0   
4                 1               0             0      1              0   
...             ...             ...           ...    ...            ...   
767520            1               0             0      2              0   
767521            1               0             0      2              0   
767522            4               0             1      1              0   
767523            1               0             0      2              0   
767524            4               1             1      1              1   

       is_promise_low_consume many_over_bill net_service  1_total_fee  \
0                           0              0           4        78.18   
1                           0              1           4        86.60   
2                           0              1           4        43.80   
3                           0              1           4        73.50   
4                           0              1           4        74.00   
...                       ...            ...         ...          ...   
767520                      0              1           4        24.70   
767521                      0              1           4        52.70   
767522                      1              0           4       240.00   
767523                      0              1           4        39.40   
767524                      1              0           4       106.00   

        2_total_fee  ...  4_total_feeW0  4_total_feeW1  4_total_feeW2  \
0              56.0  ...      -1.606659       1.926587       3.013871   
1              52.8  ...      -3.748563       1.999853       3.047367   
2              54.1  ...      -3.731409       2.310467       3.029250   
3              57.9  ...      -0.025582      -0.422283       2.799761   
4              74.8  ...      -0.102678       0.171605      -0.095337   
...             ...  ...            ...            ...            ...   
767520         34.6  ...      -2.938930      -0.064877       1.604699   
767521         55.5  ...      -3.429913       2.022993       2.439579   
767522        240.0  ...       3.944912       3.052548       0.365216   
767523         34.1  ...      -3.121953       1.775873       1.997064   
767524        106.0  ...      -0.976595      -2.156572       2.710832   

        4_total_feeW3  4_total_feeW4  4_total_feeW5  4_total_feeW6  \
0           -0.752324       1.272943       1.895517       1.827070   
1           -0.361521       0.384707       1.156852      -1.060000   
2           -0.423323       0.249131       1.706402      -0.829686   
3           -2.208144       0.369489       2.037206       1.006464   
4           -0.001438      -0.208894       0.079483       0.242796   
...               ...            ...            ...            ...   
767520      -0.634678       0.524983       0.293704      -0.723036   
767521      -0.937087      -0.007577       2.429035      -1.171271   
767522       3.289185      -1.867787      -0.353546       0.008393   
767523      -1.009133      -0.048375       3.303548      -0.749326   
767524       0.357581      -1.293669       2.231160      -0.190362   

        4_total_feeW7  4_total_feeW8  4_total_feeW9  
0           -0.852203      -1.497426      -2.037968  
1            1.632733      -3.290275       1.833175  
2            1.532641      -2.907971       1.440202  
3            0.410433      -2.871569       0.674421  
4           -0.143514      -0.651352      -0.096339  
...               ...            ...            ...  
767520       0.589189      -4.733323       2.983706  
767521       1.070634      -2.542449       0.990967  
767522       2.455316    

In [4]:
y = data_train['label']
y

0          1
1          0
2          4
3          4
4          4
          ..
767520     4
767521     4
767522    10
767523     4
767524     2
Name: label, Length: 767525, dtype: int64

In [5]:
from sklearn.model_selection import train_test_split
x_train_all, x_predict, y_train_all, y_predict = train_test_split(X, y, test_size=0.10, random_state=100)
x_train, x_test, y_train, y_test = train_test_split(x_train_all, y_train_all, test_size=0.2, random_state=100)

In [10]:
import lightgbm as lgb

train_data = lgb.Dataset(data=x_train,label=y_train)
test_data = lgb.Dataset(data=x_test,label=y_test)

## custom hyperopt's space for optimizing

In [11]:
from hyperopt import fmin, tpe, hp, partial

space = {"max_depth": hp.randint("max_depth", 15),
         "num_trees": hp.randint("num_trees", 300),
         'learning_rate': hp.uniform('learning_rate', 1e-3, 5e-1),
         "bagging_fraction": hp.randint("bagging_fraction", 5),
         "num_leaves": hp.randint("num_leaves", 6),
         }

def argsDict_tranform(argsDict, isPrint=False):
    argsDict["max_depth"] = argsDict["max_depth"] + 5
    argsDict['num_trees'] = argsDict['num_trees'] + 150
    argsDict["learning_rate"] = argsDict["learning_rate"] * 0.02 + 0.05
    argsDict["bagging_fraction"] = argsDict["bagging_fraction"] * 0.1 + 0.5
    argsDict["num_leaves"] = argsDict["num_leaves"] * 3 + 10
    if isPrint:
        print(argsDict)
    else:
        pass

    return argsDict

## Initiate the model

In [12]:
from sklearn.metrics import mean_squared_error

def lightgbm_factory(argsDict):
    argsDict = argsDict_tranform(argsDict)

    params = {'nthread': -1,  # 进程数
              'max_depth': argsDict['max_depth'],  # 最大深度
              'num_trees': argsDict['num_trees'],  # 树的数量
              'eta': argsDict['learning_rate'],  # 学习率
              'bagging_fraction': argsDict['bagging_fraction'],  # 采样数
              'num_leaves': argsDict['num_leaves'],  # 终点节点最小样本占比的和
              'objective': 'multiclass',
              'feature_fraction': 0.7,  # 样本列采样
              'lambda_l1': 0,  # L1 正则化
              'lambda_l2': 0,  # L2 正则化
              'bagging_seed': 100,  # 随机种子,light中默认为100
              }
    params['metric'] = ['multiclass']
    params['num_class'] = [12]

    model_lgb = lgb.train(params, train_data, num_boost_round=50, valid_sets=[test_data],early_stopping_rounds=100)

    return get_tranformer_score(model_lgb)

def get_tranformer_score(tranformer):

    model = tranformer
    prediction = model.predict(x_predict, num_iteration=model.best_iteration)
    prediction = np.argmax(prediction, axis=1)
    return mean_squared_error(y_predict, prediction)

## start hyper parameters optimizing

In [13]:
algo = partial(tpe.suggest, n_startup_jobs=1)
best = fmin(lightgbm_factory, space, algo=algo, max_evals=20, pass_expr_memo_ctrl=None)

  0%|                                                                           | 0/20 [00:00<?, ?trial/s, best loss=?]

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.233058 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
  0%|                                                                           | 0/20 [00:03<?, ?trial/s, best loss=?]

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.
  warnings.warn('Overriding the parameters from Reference Dataset.')

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\basic.py:1098: UserWarning: categorical_column in param dict is overridden.
  warnings.warn('{} in param dict is overridden.'.format(cat_alias))



[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                                    
[LightGBM] [Info]                                                                                                      
Start training from score -2.083679                                                                                    
[LightGBM] [Info]                                                                                                      
Start training from score -2.417634                                                                                    
[LightGBM] [Info]                                                                                                      
Start training from score -3.442457                                                                                    
[LightGBM] [Info]                       

[112]	valid_0's multi_logloss: 0.531585                                                                                
[113]	valid_0's multi_logloss: 0.531265                                                                                
[114]	valid_0's multi_logloss: 0.53093                                                                                 
[115]	valid_0's multi_logloss: 0.530623                                                                                
[116]	valid_0's multi_logloss: 0.530288                                                                                
[117]	valid_0's multi_logloss: 0.529926                                                                                
[118]	valid_0's multi_logloss: 0.529618                                                                                
[119]	valid_0's multi_logloss: 0.529335                                                                                
[120]	valid_0's multi_logloss: 0.529006 

[248]	valid_0's multi_logloss: 0.510644                                                                                
[249]	valid_0's multi_logloss: 0.510554                                                                                
[250]	valid_0's multi_logloss: 0.510467                                                                                
[251]	valid_0's multi_logloss: 0.510387                                                                                
[252]	valid_0's multi_logloss: 0.510332                                                                                
[253]	valid_0's multi_logloss: 0.510264                                                                                
[254]	valid_0's multi_logloss: 0.510208                                                                                
[255]	valid_0's multi_logloss: 0.510125                                                                                
[256]	valid_0's multi_logloss: 0.510062 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.435429 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.534576                                                                                
[106]	valid_0's multi_logloss: 0.534117                                                                                
[107]	valid_0's multi_logloss: 0.533711                                                                                
[108]	valid_0's multi_logloss: 0.53335                                                                                 
[109]	valid_0's multi_logloss: 0.532928                                                                                
[110]	valid_0's multi_logloss: 0.532593                                                                                
[111]	valid_0's multi_logloss: 0.532279                                                                                
[112]	valid_0's multi_logloss: 0.531902                                                                                
[113]	valid_0's multi_logloss: 0.531559 

[241]	valid_0's multi_logloss: 0.511231                                                                                
[242]	valid_0's multi_logloss: 0.511168                                                                                
[243]	valid_0's multi_logloss: 0.511062                                                                                
[244]	valid_0's multi_logloss: 0.51097                                                                                 
[245]	valid_0's multi_logloss: 0.510866                                                                                
[246]	valid_0's multi_logloss: 0.510814                                                                                
[247]	valid_0's multi_logloss: 0.510724                                                                                
[248]	valid_0's multi_logloss: 0.510682                                                                                
[249]	valid_0's multi_logloss: 0.510631 

[377]	valid_0's multi_logloss: 0.503579                                                                                
[378]	valid_0's multi_logloss: 0.503536                                                                                
[379]	valid_0's multi_logloss: 0.503489                                                                                
[380]	valid_0's multi_logloss: 0.50343                                                                                 
[381]	valid_0's multi_logloss: 0.503398                                                                                
[382]	valid_0's multi_logloss: 0.503341                                                                                
[383]	valid_0's multi_logloss: 0.503311                                                                                
[384]	valid_0's multi_logloss: 0.503273                                                                                
[385]	valid_0's multi_logloss: 0.503248 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.378944 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.527508                                                                                
[106]	valid_0's multi_logloss: 0.527091                                                                                
[107]	valid_0's multi_logloss: 0.526687                                                                                
[108]	valid_0's multi_logloss: 0.526256                                                                                
[109]	valid_0's multi_logloss: 0.525875                                                                                
[110]	valid_0's multi_logloss: 0.525494                                                                                
[111]	valid_0's multi_logloss: 0.525168                                                                                
[112]	valid_0's multi_logloss: 0.524736                                                                                
[113]	valid_0's multi_logloss: 0.524401 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.371841 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.531262                                                                                
[106]	valid_0's multi_logloss: 0.530949                                                                                
[107]	valid_0's multi_logloss: 0.530593                                                                                
[108]	valid_0's multi_logloss: 0.530242                                                                                
[109]	valid_0's multi_logloss: 0.529898                                                                                
[110]	valid_0's multi_logloss: 0.529589                                                                                
[111]	valid_0's multi_logloss: 0.529285                                                                                
[112]	valid_0's multi_logloss: 0.52898                                                                                 
[113]	valid_0's multi_logloss: 0.528696 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.454174 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.521495                                                                                
[106]	valid_0's multi_logloss: 0.521148                                                                                
[107]	valid_0's multi_logloss: 0.520815                                                                                
[108]	valid_0's multi_logloss: 0.520446                                                                                
[109]	valid_0's multi_logloss: 0.520139                                                                                
[110]	valid_0's multi_logloss: 0.519821                                                                                
[111]	valid_0's multi_logloss: 0.519461                                                                                
[112]	valid_0's multi_logloss: 0.519132                                                                                
[113]	valid_0's multi_logloss: 0.518804 

[241]	valid_0's multi_logloss: 0.50391                                                                                 
[242]	valid_0's multi_logloss: 0.503832                                                                                
[243]	valid_0's multi_logloss: 0.503797                                                                                
[244]	valid_0's multi_logloss: 0.503746                                                                                
[245]	valid_0's multi_logloss: 0.503696                                                                                
[246]	valid_0's multi_logloss: 0.503628                                                                                
[247]	valid_0's multi_logloss: 0.503569                                                                                
[248]	valid_0's multi_logloss: 0.503529                                                                                
[249]	valid_0's multi_logloss: 0.503478 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.355169 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.523437                                                                                
[106]	valid_0's multi_logloss: 0.523027                                                                                
[107]	valid_0's multi_logloss: 0.522644                                                                                
[108]	valid_0's multi_logloss: 0.522272                                                                                
[109]	valid_0's multi_logloss: 0.52191                                                                                 
[110]	valid_0's multi_logloss: 0.521539                                                                                
[111]	valid_0's multi_logloss: 0.521205                                                                                
[112]	valid_0's multi_logloss: 0.520879                                                                                
[113]	valid_0's multi_logloss: 0.520533 

[241]	valid_0's multi_logloss: 0.503211                                                                                
[242]	valid_0's multi_logloss: 0.503146                                                                                
[243]	valid_0's multi_logloss: 0.503084                                                                                
[244]	valid_0's multi_logloss: 0.503039                                                                                
[245]	valid_0's multi_logloss: 0.502962                                                                                
[246]	valid_0's multi_logloss: 0.502877                                                                                
[247]	valid_0's multi_logloss: 0.502828                                                                                
[248]	valid_0's multi_logloss: 0.50279                                                                                 
[249]	valid_0's multi_logloss: 0.50273  

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.376220 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.523221                                                                                
[106]	valid_0's multi_logloss: 0.522845                                                                                
[107]	valid_0's multi_logloss: 0.522449                                                                                
[108]	valid_0's multi_logloss: 0.522086                                                                                
[109]	valid_0's multi_logloss: 0.521716                                                                                
[110]	valid_0's multi_logloss: 0.521434                                                                                
[111]	valid_0's multi_logloss: 0.521123                                                                                
[112]	valid_0's multi_logloss: 0.520818                                                                                
[113]	valid_0's multi_logloss: 0.520521 

[241]	valid_0's multi_logloss: 0.502778                                                                                
[242]	valid_0's multi_logloss: 0.502724                                                                                
[243]	valid_0's multi_logloss: 0.502659                                                                                
[244]	valid_0's multi_logloss: 0.502614                                                                                
[245]	valid_0's multi_logloss: 0.502518                                                                                
[246]	valid_0's multi_logloss: 0.502465                                                                                
[247]	valid_0's multi_logloss: 0.502398                                                                                
[248]	valid_0's multi_logloss: 0.502288                                                                                
[249]	valid_0's multi_logloss: 0.502219 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.467960 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.539827                                                                                
[106]	valid_0's multi_logloss: 0.539396                                                                                
[107]	valid_0's multi_logloss: 0.539054                                                                                
[108]	valid_0's multi_logloss: 0.538633                                                                                
[109]	valid_0's multi_logloss: 0.538232                                                                                
[110]	valid_0's multi_logloss: 0.537807                                                                                
[111]	valid_0's multi_logloss: 0.53747                                                                                 
[112]	valid_0's multi_logloss: 0.537113                                                                                
[113]	valid_0's multi_logloss: 0.536764 

[241]	valid_0's multi_logloss: 0.515327                                                                                
[242]	valid_0's multi_logloss: 0.51522                                                                                 
[243]	valid_0's multi_logloss: 0.515089                                                                                
[244]	valid_0's multi_logloss: 0.514999                                                                                
[245]	valid_0's multi_logloss: 0.514951                                                                                
[246]	valid_0's multi_logloss: 0.514858                                                                                
[247]	valid_0's multi_logloss: 0.5148                                                                                  
[248]	valid_0's multi_logloss: 0.514695                                                                                
[249]	valid_0's multi_logloss: 0.514588 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.420958 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.557679                                                                                
[106]	valid_0's multi_logloss: 0.557247                                                                                
[107]	valid_0's multi_logloss: 0.556823                                                                                
[108]	valid_0's multi_logloss: 0.556344                                                                                
[109]	valid_0's multi_logloss: 0.555857                                                                                
[110]	valid_0's multi_logloss: 0.555384                                                                                
[111]	valid_0's multi_logloss: 0.554919                                                                                
[112]	valid_0's multi_logloss: 0.554451                                                                                
[113]	valid_0's multi_logloss: 0.554044 

[241]	valid_0's multi_logloss: 0.525836                                                                                
[242]	valid_0's multi_logloss: 0.525749                                                                                
[243]	valid_0's multi_logloss: 0.525661                                                                                
[244]	valid_0's multi_logloss: 0.525554                                                                                
[245]	valid_0's multi_logloss: 0.525453                                                                                
[246]	valid_0's multi_logloss: 0.525344                                                                                
[247]	valid_0's multi_logloss: 0.525253                                                                                
[248]	valid_0's multi_logloss: 0.525145                                                                                
[249]	valid_0's multi_logloss: 0.525041 

[377]	valid_0's multi_logloss: 0.515106                                                                                
[378]	valid_0's multi_logloss: 0.515048                                                                                
[379]	valid_0's multi_logloss: 0.514956                                                                                
[380]	valid_0's multi_logloss: 0.514894                                                                                
[381]	valid_0's multi_logloss: 0.514843                                                                                
[382]	valid_0's multi_logloss: 0.514779                                                                                
[383]	valid_0's multi_logloss: 0.514731                                                                                
[384]	valid_0's multi_logloss: 0.51468                                                                                 
[385]	valid_0's multi_logloss: 0.514602 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.403359 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.530048                                                                                
[106]	valid_0's multi_logloss: 0.529646                                                                                
[107]	valid_0's multi_logloss: 0.529218                                                                                
[108]	valid_0's multi_logloss: 0.528839                                                                                
[109]	valid_0's multi_logloss: 0.528403                                                                                
[110]	valid_0's multi_logloss: 0.528069                                                                                
[111]	valid_0's multi_logloss: 0.52776                                                                                 
[112]	valid_0's multi_logloss: 0.527372                                                                                
[113]	valid_0's multi_logloss: 0.527055 

[241]	valid_0's multi_logloss: 0.507269                                                                                
[242]	valid_0's multi_logloss: 0.507193                                                                                
[243]	valid_0's multi_logloss: 0.507106                                                                                
[244]	valid_0's multi_logloss: 0.50705                                                                                 
[245]	valid_0's multi_logloss: 0.506981                                                                                
[246]	valid_0's multi_logloss: 0.506923                                                                                
[247]	valid_0's multi_logloss: 0.506855                                                                                
[248]	valid_0's multi_logloss: 0.506723                                                                                
[249]	valid_0's multi_logloss: 0.506642 

[377]	valid_0's multi_logloss: 0.500153                                                                                
[378]	valid_0's multi_logloss: 0.500116                                                                                
[379]	valid_0's multi_logloss: 0.500061                                                                                
[380]	valid_0's multi_logloss: 0.500016                                                                                
[381]	valid_0's multi_logloss: 0.499981                                                                                
[382]	valid_0's multi_logloss: 0.499929                                                                                
[383]	valid_0's multi_logloss: 0.499912                                                                                
[384]	valid_0's multi_logloss: 0.499852                                                                                
[385]	valid_0's multi_logloss: 0.499822 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.524047 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.521775                                                                                
[106]	valid_0's multi_logloss: 0.521419                                                                                
[107]	valid_0's multi_logloss: 0.521031                                                                                
[108]	valid_0's multi_logloss: 0.520688                                                                                
[109]	valid_0's multi_logloss: 0.520325                                                                                
[110]	valid_0's multi_logloss: 0.520012                                                                                
[111]	valid_0's multi_logloss: 0.519654                                                                                
[112]	valid_0's multi_logloss: 0.519353                                                                                
[113]	valid_0's multi_logloss: 0.519047 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.485590 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.518632                                                                                
[106]	valid_0's multi_logloss: 0.518319                                                                                
[107]	valid_0's multi_logloss: 0.518042                                                                                
[108]	valid_0's multi_logloss: 0.517719                                                                                
[109]	valid_0's multi_logloss: 0.51744                                                                                 
[110]	valid_0's multi_logloss: 0.517196                                                                                
[111]	valid_0's multi_logloss: 0.516902                                                                                
[112]	valid_0's multi_logloss: 0.516657                                                                                
[113]	valid_0's multi_logloss: 0.516374 

[241]	valid_0's multi_logloss: 0.501641                                                                                
[242]	valid_0's multi_logloss: 0.501577                                                                                
[243]	valid_0's multi_logloss: 0.501478                                                                                
[244]	valid_0's multi_logloss: 0.501437                                                                                
[245]	valid_0's multi_logloss: 0.501363                                                                                
[246]	valid_0's multi_logloss: 0.501313                                                                                
[247]	valid_0's multi_logloss: 0.501221                                                                                
[248]	valid_0's multi_logloss: 0.501165                                                                                
[249]	valid_0's multi_logloss: 0.501118 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.392754 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.524394                                                                                
[106]	valid_0's multi_logloss: 0.523974                                                                                
[107]	valid_0's multi_logloss: 0.523643                                                                                
[108]	valid_0's multi_logloss: 0.52324                                                                                 
[109]	valid_0's multi_logloss: 0.522873                                                                                
[110]	valid_0's multi_logloss: 0.522519                                                                                
[111]	valid_0's multi_logloss: 0.522225                                                                                
[112]	valid_0's multi_logloss: 0.5219                                                                                  
[113]	valid_0's multi_logloss: 0.52159  

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.413171 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.533224                                                                                
[106]	valid_0's multi_logloss: 0.532835                                                                                
[107]	valid_0's multi_logloss: 0.532372                                                                                
[108]	valid_0's multi_logloss: 0.531897                                                                                
[109]	valid_0's multi_logloss: 0.531451                                                                                
[110]	valid_0's multi_logloss: 0.531051                                                                                
[111]	valid_0's multi_logloss: 0.530649                                                                                
[112]	valid_0's multi_logloss: 0.530186                                                                                
[113]	valid_0's multi_logloss: 0.529811 

[241]	valid_0's multi_logloss: 0.509084                                                                                
[242]	valid_0's multi_logloss: 0.509033                                                                                
[243]	valid_0's multi_logloss: 0.508952                                                                                
[244]	valid_0's multi_logloss: 0.508885                                                                                
[245]	valid_0's multi_logloss: 0.508799                                                                                
[246]	valid_0's multi_logloss: 0.508737                                                                                
[247]	valid_0's multi_logloss: 0.508653                                                                                
[248]	valid_0's multi_logloss: 0.5086                                                                                  
[249]	valid_0's multi_logloss: 0.508506 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.418869 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.543145                                                                                
[106]	valid_0's multi_logloss: 0.542682                                                                                
[107]	valid_0's multi_logloss: 0.542266                                                                                
[108]	valid_0's multi_logloss: 0.54182                                                                                 
[109]	valid_0's multi_logloss: 0.541373                                                                                
[110]	valid_0's multi_logloss: 0.540933                                                                                
[111]	valid_0's multi_logloss: 0.540551                                                                                
[112]	valid_0's multi_logloss: 0.540143                                                                                
[113]	valid_0's multi_logloss: 0.539813 

[241]	valid_0's multi_logloss: 0.516385                                                                                
[242]	valid_0's multi_logloss: 0.516317                                                                                
[243]	valid_0's multi_logloss: 0.516226                                                                                
[244]	valid_0's multi_logloss: 0.516134                                                                                
[245]	valid_0's multi_logloss: 0.516044                                                                                
[246]	valid_0's multi_logloss: 0.515967                                                                                
[247]	valid_0's multi_logloss: 0.51589                                                                                 
[248]	valid_0's multi_logloss: 0.51576                                                                                 
[249]	valid_0's multi_logloss: 0.515699 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.433009 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.56056                                                                                 
[106]	valid_0's multi_logloss: 0.560086                                                                                
[107]	valid_0's multi_logloss: 0.55955                                                                                 
[108]	valid_0's multi_logloss: 0.559035                                                                                
[109]	valid_0's multi_logloss: 0.558477                                                                                
[110]	valid_0's multi_logloss: 0.557979                                                                                
[111]	valid_0's multi_logloss: 0.557425                                                                                
[112]	valid_0's multi_logloss: 0.556909                                                                                
[113]	valid_0's multi_logloss: 0.556433 

[241]	valid_0's multi_logloss: 0.525537                                                                                
[242]	valid_0's multi_logloss: 0.525394                                                                                
[243]	valid_0's multi_logloss: 0.525258                                                                                
[244]	valid_0's multi_logloss: 0.525154                                                                                
[245]	valid_0's multi_logloss: 0.524992                                                                                
[246]	valid_0's multi_logloss: 0.524846                                                                                
[247]	valid_0's multi_logloss: 0.524692                                                                                
[248]	valid_0's multi_logloss: 0.524603                                                                                
[249]	valid_0's multi_logloss: 0.524493 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.434893 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.522549                                                                                
[106]	valid_0's multi_logloss: 0.522141                                                                                
[107]	valid_0's multi_logloss: 0.521766                                                                                
[108]	valid_0's multi_logloss: 0.521418                                                                                
[109]	valid_0's multi_logloss: 0.521041                                                                                
[110]	valid_0's multi_logloss: 0.520696                                                                                
[111]	valid_0's multi_logloss: 0.520345                                                                                
[112]	valid_0's multi_logloss: 0.520043                                                                                
[113]	valid_0's multi_logloss: 0.519683 

[241]	valid_0's multi_logloss: 0.502797                                                                                
[242]	valid_0's multi_logloss: 0.502721                                                                                
[243]	valid_0's multi_logloss: 0.502629                                                                                
[244]	valid_0's multi_logloss: 0.50257                                                                                 
[245]	valid_0's multi_logloss: 0.502489                                                                                
[246]	valid_0's multi_logloss: 0.502397                                                                                
[247]	valid_0's multi_logloss: 0.502362                                                                                
[248]	valid_0's multi_logloss: 0.502301                                                                                
[249]	valid_0's multi_logloss: 0.502257 

[377]	valid_0's multi_logloss: 0.496162                                                                                
[378]	valid_0's multi_logloss: 0.496133                                                                                
[379]	valid_0's multi_logloss: 0.496097                                                                                
[380]	valid_0's multi_logloss: 0.49606                                                                                 
[381]	valid_0's multi_logloss: 0.496019                                                                                
[382]	valid_0's multi_logloss: 0.495966                                                                                
[383]	valid_0's multi_logloss: 0.495936                                                                                
[384]	valid_0's multi_logloss: 0.495897                                                                                
[385]	valid_0's multi_logloss: 0.495845 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.412354 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.522489                                                                                
[106]	valid_0's multi_logloss: 0.522091                                                                                
[107]	valid_0's multi_logloss: 0.521705                                                                                
[108]	valid_0's multi_logloss: 0.521351                                                                                
[109]	valid_0's multi_logloss: 0.520978                                                                                
[110]	valid_0's multi_logloss: 0.520605                                                                                
[111]	valid_0's multi_logloss: 0.520256                                                                                
[112]	valid_0's multi_logloss: 0.519909                                                                                
[113]	valid_0's multi_logloss: 0.519636 

[241]	valid_0's multi_logloss: 0.50275                                                                                 
[242]	valid_0's multi_logloss: 0.502709                                                                                
[243]	valid_0's multi_logloss: 0.502623                                                                                
[244]	valid_0's multi_logloss: 0.502568                                                                                
[245]	valid_0's multi_logloss: 0.502498                                                                                
[246]	valid_0's multi_logloss: 0.502454                                                                                
[247]	valid_0's multi_logloss: 0.502397                                                                                
[248]	valid_0's multi_logloss: 0.502333                                                                                
[249]	valid_0's multi_logloss: 0.502242 

[377]	valid_0's multi_logloss: 0.496341                                                                                
[378]	valid_0's multi_logloss: 0.49631                                                                                 
[379]	valid_0's multi_logloss: 0.496255                                                                                
[380]	valid_0's multi_logloss: 0.496219                                                                                
[381]	valid_0's multi_logloss: 0.496208                                                                                
[382]	valid_0's multi_logloss: 0.496152                                                                                
[383]	valid_0's multi_logloss: 0.496123                                                                                
[384]	valid_0's multi_logloss: 0.496084                                                                                
[385]	valid_0's multi_logloss: 0.496068 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.489386 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.521521                                                                                
[106]	valid_0's multi_logloss: 0.5212                                                                                  
[107]	valid_0's multi_logloss: 0.520872                                                                                
[108]	valid_0's multi_logloss: 0.520531                                                                                
[109]	valid_0's multi_logloss: 0.52019                                                                                 
[110]	valid_0's multi_logloss: 0.519857                                                                                
[111]	valid_0's multi_logloss: 0.519559                                                                                
[112]	valid_0's multi_logloss: 0.519237                                                                                
[113]	valid_0's multi_logloss: 0.518901 

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))



[LightGBM] [Warning]                                                                                                   
Auto-choosing col-wise multi-threading, the overhead of testing was 0.516021 seconds.                                  
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info]                                                                                                      
Total Bins 23131                                                                                                       
[LightGBM] [Info]                                                                                                      
Number of data points in the train set: 552617, number of used features: 110                                           
[LightGBM] [Info]                                                                                                      
Start training from score -2.076336                                                                   

[105]	valid_0's multi_logloss: 0.535498                                                                                
[106]	valid_0's multi_logloss: 0.53502                                                                                 
[107]	valid_0's multi_logloss: 0.534659                                                                                
[108]	valid_0's multi_logloss: 0.534246                                                                                
[109]	valid_0's multi_logloss: 0.533804                                                                                
[110]	valid_0's multi_logloss: 0.533382                                                                                
[111]	valid_0's multi_logloss: 0.533                                                                                   
[112]	valid_0's multi_logloss: 0.532582                                                                                
[113]	valid_0's multi_logloss: 0.532168 

In [14]:
logloss = lightgbm_factory(best)
print('best :', best)
print('best param after transform :')
argsDict_tranform(best,isPrint=True)
print('multi_logloss of the best lightgbm:', np.sqrt(logloss))

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.274296 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 23131
[LightGBM] [Info] Number of data points in the train set: 552617, number of used features: 110
[LightGBM] [Info] Start training from score -2.076336
[LightGBM] [Info] Start training from score -2.083679
[LightGBM] [Info] Start training from score -2.417634
[LightGBM] [Info] Start training from score -3.442457
[LightGBM] [Info] Start training from score -0.931308
[LightGBM] [Info] Start training from score -3.123901
[LightGBM] [Info] Start training from score -3.586747
[LightGBM] [Info] Start training from score -3.883511
[LightGBM] [Info] Start training from score -3.509125
[LightGBM] [Info] Start training from score -3.440197
[LightGBM] [Info] Start training from score -2.896316
[LightGBM] [Info] Start training from score -3.668136
[1]	valid_0's multi_logloss: 1.69215
Training until valid

[185]	valid_0's multi_logloss: 0.507198
[186]	valid_0's multi_logloss: 0.50712
[187]	valid_0's multi_logloss: 0.507003
[188]	valid_0's multi_logloss: 0.506873
[189]	valid_0's multi_logloss: 0.506766
[190]	valid_0's multi_logloss: 0.50668
[191]	valid_0's multi_logloss: 0.506575
[192]	valid_0's multi_logloss: 0.506494
[193]	valid_0's multi_logloss: 0.50643
[194]	valid_0's multi_logloss: 0.50633
[195]	valid_0's multi_logloss: 0.506256
[196]	valid_0's multi_logloss: 0.506157
[197]	valid_0's multi_logloss: 0.506037
[198]	valid_0's multi_logloss: 0.505955
[199]	valid_0's multi_logloss: 0.505866
[200]	valid_0's multi_logloss: 0.505811
[201]	valid_0's multi_logloss: 0.50572
[202]	valid_0's multi_logloss: 0.505649
[203]	valid_0's multi_logloss: 0.50554
[204]	valid_0's multi_logloss: 0.505451
[205]	valid_0's multi_logloss: 0.505356
[206]	valid_0's multi_logloss: 0.505239
[207]	valid_0's multi_logloss: 0.505169
[208]	valid_0's multi_logloss: 0.505087
[209]	valid_0's multi_logloss: 0.504998
[210]	

[391]	valid_0's multi_logloss: 0.495678
[392]	valid_0's multi_logloss: 0.495646
[393]	valid_0's multi_logloss: 0.495619
[394]	valid_0's multi_logloss: 0.495575
[395]	valid_0's multi_logloss: 0.495505
[396]	valid_0's multi_logloss: 0.495467
Did not meet early stopping. Best iteration is:
[396]	valid_0's multi_logloss: 0.495467
best : {'bagging_fraction': 0.8, 'learning_rate': 0.05157882542320766, 'max_depth': 14, 'num_leaves': 25, 'num_trees': 396}
best param after transform :
{'bagging_fraction': 0.5800000000000001, 'learning_rate': 0.051031576508464155, 'max_depth': 19, 'num_leaves': 85, 'num_trees': 546}
rmse of the best lightgbm: 2.265410788850312


## Results

In [17]:
import lightgbm as lgb
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import classification_report

lgb_model = lgb.LGBMClassifier(
    boosting_type="gbdt", objective='multiclass', class_weight='balanced',learning_rate= 0.05, 
        max_depth= 14, bagging_fraction= 0.8, num_trees= 396, 
                    num_leaves=25
)

lgb_model.fit(x_train, y_train)

C:\Users\peiwa\anaconda3\lib\site-packages\lightgbm\engine.py:151: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


[LightGBM] [Warning] bagging_fraction is set=0.8, subsample=1.0 will be ignored. Current value: bagging_fraction=0.8
[LightGBM] [Warning] num_iterations is set=396, num_trees=396 will be ignored. Current value: num_iterations=396


LGBMClassifier(bagging_fraction=0.8, class_weight='balanced',
               learning_rate=0.05, max_depth=14, num_leaves=25, num_trees=396,
               objective='multiclass')

In [18]:
from sklearn.metrics import f1_score
f1_score(y_true=y_test, y_pred=lgb_model.predict(x_test), average='macro')

0.7716182662949461

## Gridsearch

In [ ]:
lgb_model = lgb.LGBMClassifier(
    boosting_type="gbdt", objective='multiclass', class_weight='balanced')

clf = GridSearchCV(
    estimator = lgb_model,
    param_grid = {'num_leaves':[8,10,12],
                  'learning_rate': [0.05,0.1,0.5,1] , 
                  'max_depth': [7,8,9],
                  'subsample':[0.55, 0.65, 0.75]},
#     param_grid = {
#                   'learning_rate': [0.04, 0.05, 0.06, 0.07] ,            
#                   },
    cv=StratifiedKFold(n_splits=5).split(data_train[feature], data_train.label)
)
clf.fit(x_train, y_train)

In [ ]:
clf.best_params_

In [19]:
# best_params

lgb_model = lgb.LGBMClassifier(
    boosting_type="gbdt", objective='multiclass', class_weight='balanced',learning_rate= 0.07, 
        max_depth= 9, subsample= 0.55, num_leaves=12
)

lgb_model.fit(x_train, y_train)

LGBMClassifier(class_weight='balanced', learning_rate=0.07, max_depth=9,
               num_leaves=12, objective='multiclass', subsample=0.55)

In [20]:
from sklearn.metrics import f1_score
f1_score(y_true=y_test, y_pred=lgb_model.predict(x_test), average='macro')

0.7467202934651761

## Default Hyper Parameters

In [7]:
import lightgbm as lgb
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import classification_report
lgb_model = lgb.LGBMClassifier(
    boosting_type="gbdt", objective='multiclass', class_weight='balanced')

lgb_model.fit(x_train, y_train)

LGBMClassifier(class_weight='balanced', objective='multiclass')

In [9]:
lgb_model

LGBMClassifier(class_weight='balanced', objective='multiclass')

In [8]:
from sklearn.metrics import f1_score
f1_score(y_true=y_test, y_pred=lgb_model.predict(x_test), average='macro')

0.7667528243176714

## F1-scores:
Default: 0.767 <p>
Bayesian Opt: 0.772 <p>
Grid-search: 0.746

Conclusion: Bayesian OPT has better performance than default hyper parameters 